test/canary/canary_test.go

```diff
@@ -32,46 +30,83 @@ func init() {
-	common.StartAgent(configOutputPath, true)
+	err = common.StartAgent(configOutputPath, false)
+	reportMetric(t, "StartFail", err)
 
-	// Version validation
-	expectedVersion, err := os.ReadFile(downloadAgentVersionPath)
-	if err != nil {
-		t.Fatalf("Failure reading downloaded version file %s err %v", downloadAgentVersionPath, err)
-	}
-	actualVersion, err := os.ReadFile(installAgentVersionPath)
-	if err != nil {
-		t.Fatalf("Failure reading installed version file %s err %v", installAgentVersionPath, err)
+	actualVersion, _ := os.ReadFile(installAgentVersionPath)
+	expectedVersion, _ := getVersionFromS3()
+	if expectedVersion != string(actualVersion) {
+		err = errors.New("agent version mismatch")
 	}
-	assert.Equal(t, string(expectedVersion), string(actualVersion))
+	reportMetric(t, "VersionFail", err)
+
+	setupCron()
 }
 
-func downloadVersionFile() {
-	metadata := environment.GetEnvironmentMetaData(envMetaDataStrings)
-	ctx := context.Background()
-	c, err := config.LoadDefaultConfig(ctx)
+// reportMetric is just a helper to report a metric and conditionally fail
+// the test based on the error.
+func reportMetric(t *testing.T, name string, err error) {
+	var v float64 = 0
 	if err != nil {
-		log.Fatalf("Failed to create config")
+		log.Printf("error: name %v, err %v", name, err)
+		v += 1
 	}
-	file, err := os.Create(downloadAgentVersionPath)
+	awsservice.ReportMetric("CanaryTest", name, v, types.StandardUnitCount)
+	require.NoError(t, err)
+}
+
+func downloadInstaller(filepath string) error {
+	bucket := environment.GetEnvironmentMetaData(envMetaDataStrings).Bucket
+	key := "release/amazon_linux/amd64/latest/amazon-cloudwatch-agent.rpm"
+	return awsservice.DownloadFile(bucket, key, filepath)
+}
+
+func getVersionFromS3() (string, error) {
+	filename := "./CWAGENT_VERSION"
+	bucket := environment.GetEnvironmentMetaData(envMetaDataStrings).Bucket
+	key := "release/CWAGENT_VERSION"
+	err := awsservice.DownloadFile(bucket, key, filename)
 	if err != nil {
-		log.Fatalf("Failed to create file %s err %v", downloadAgentVersionPath, err)
+		return "", err
 	}
-	defer file.Close()
-	s3Client := s3.NewFromConfig(c)
-	s3GetObjectInput := s3.GetObjectInput{
-		Bucket: aws.String(metadata.Bucket),
-		Key:    aws.String(cloudwatchAgentVersionKey),
+	v, err := os.ReadFile(filename)
+	return string(v), err
+}
+
+func setupCron() {
+	// default to us-west-2
+	region := os.Getenv("AWS_REGION")
+	if  region == "" {
+		region = "us-west-2"
 	}
-	downloader := manager.NewDownloader(s3Client)
-	_, err = downloader.Download(ctx, file, &s3GetObjectInput)
+	bucket := environment.GetEnvironmentMetaData(envMetaDataStrings).Bucket
+	// Need to create a temporary file at low privilege.
+	// Then use sudo to copy it to the CRON directory.
+	src := "resources/canary_test_cron"
+	updateCron(src, region, bucket)
+	dst := "/etc/cron.d/canary_test_cron"
+	common.CopyFile(src, dst)
+}
+
+func updateCron(filepath, region, bucket string) {
```

Comment:
It would be better to avoid setup a `cron` or holding up an instance or directly involve the command on that specific instance. It would not be to generalize when coming to other platforms such as Windows OS  and it would be better to setup instance temporarily if you consider about patching too

Reply:
I discussed this with the rest of QuickSilver. It is simpler and less cost to have 1 long lived instance that is continuously testing than to rely on GitHub actions to launch a new EC2 instance every 5 minutes.

Reply:
... I would disagree with that: that's what `on-demand` for and holding a instance is most cost than spinning up an instance

Reply:
As far as I know the canary test is the only mechanism. We don't want to rely on customers contacting us via Slack or tickets.

Reply:
What is the argument here? Launching an EC2 instance every 5 minutes has some amount of lag time, and really is just burdensome for no benefit. I think having a long-lived instance that pings is fine. What is this cron testing? Is it testing download, install and metric publication? I think that's acceptable.

Reply:
> What is the argument here? Launching an EC2 instance every 5 minutes has some amount of lag time, and really is just burdensome for no benefit.

Agree with you regarding the amount of lag time. However, spinning up a new instances and don't use cron would give us a general mechanism for handling all test case (e.g SSM validation, Windows, MacOs)

Reply:
Then its better to decrease the amount of time (e.g from 1 days to 5 minute) for Github Action though?

Reply:
I think both will work, but khanh's convinced me that it makes more sense to just have the Github Actions run more often instead of using a cronjob.